```
E           AssertionError: assert False
E            +  where False = <built-in method startswith of str object at 0xee344710>('| Iterator SizeOf: 488')
E            +    where <built-in method startswith of str object at 0xee344710> = '| Iterator SizeOf: 264'.startswith

```

--- a/numpy/core/src/multiarray/nditer_pywrap.c
+++ b/numpy/core/src/multiarray/nditer_pywrap.c
@@ -1134,6 +1134,9 @@ static void
 npyiter_dealloc(NewNpyArrayIterObject *self)
 {
     if (self->iter) {
+        /* Store error, so that WriteUnraisable cannot clear an existing one */
+        PyObject *exc, *val, *tb;
+        PyErr_Fetch(&exc, &val, &tb);
         if (npyiter_has_writeback(self->iter)) {
             if (PyErr_WarnEx(PyExc_RuntimeWarning,
                     "Temporary data has not been written back to one of the "
@@ -1152,10 +1155,13 @@ npyiter_dealloc(NewNpyArrayIterObject *self)
                 }
             }
         }
-        NpyIter_Deallocate(self->iter);
+        if (!NpyIter_Deallocate(self->iter)) {
+            PyErr_WriteUnraisable(Py_None);
+        }
         self->iter = NULL;
         Py_XDECREF(self->nested_child);
         self->nested_child = NULL;
+        PyErr_Restore(exc, val, tb);
     }
     Py_TYPE(self)->tp_free((PyObject*)self);
 }
@@ -2320,7 +2326,7 @@ npyiter_close(NewNpyArrayIterObject *self)
     self->iter = NULL;
     Py_XDECREF(self->nested_child);
     self->nested_child = NULL;
-    if (ret < 0) {
+    if (ret != NPY_SUCCEED) {
         return NULL;
     }
     Py_RETURN_NONE;

--- a/numpy/core/tests/test_nditer.py
+++ b/numpy/core/tests/test_nditer.py
@@ -1,6 +1,8 @@
 import sys
 import pytest
 
+import subprocess, textwrap, re
+
 import numpy as np
 import numpy.core._multiarray_tests as _multiarray_tests
 from numpy import array, arange, nditer, all
@@ -3014,3 +3016,86 @@ def test_partial_iteration_error(in_dtype, buf_dtype):
         it.iternext()
 
     assert count == sys.getrefcount(value)
+
+
+def test_debug_print():
+    """
+    Matches the expected output of a debug print with the actual output.
+    Note that the iterator dump should not be considered stable API,
+    this test is mainly to ensure the print does not crash.
+
+    Currently uses a subprocess to avoid dealing with the C level `printf`s.
+    """
+    # the expected output with all addresses and sizes stripped (they vary
+    # and/or are platform dependend).
+    expected = textwrap.dedent("""
+    ------ BEGIN ITERATOR DUMP ------
+    | Iterator Address:
+    | ItFlags: BUFFER REDUCE REUSE_REDUCE_LOOPS
+    | NDim: 2
+    | NOp: 2
+    | IterSize: 50
+    | IterStart: 0
+    | IterEnd: 50
+    | IterIndex: 0
+    | Iterator SizeOf:
+    | BufferData SizeOf:
+    | AxisData SizeOf:
+    |
+    | Perm: 0 1
+    | DTypes:
+    | DTypes: dtype('float64') dtype('int32')
+    | InitDataPtrs:
+    | BaseOffsets: 0 0
+    | Operands:
+    | Operand DTypes: dtype('int64') dtype('float64')
+    | OpItFlags:
+    |   Flags[0]: READ CAST ALIGNED
+    |   Flags[1]: READ WRITE CAST ALIGNED REDUCE
+    |
+    | BufferData:
+    |   BufferSize: 50
+    |   Size: 5
+    |   BufIterEnd: 5
+    |   REDUCE Pos: 0
+    |   REDUCE OuterSize: 10
+    |   REDUCE OuterDim: 1
+    |   Strides: 8 4
+    |   Ptrs:
+    |   REDUCE Outer Strides: 40 0
+    |   REDUCE Outer Ptrs:
+    |   ReadTransferFn:
+    |   ReadTransferData: (nil) (nil)
+    |   WriteTransferFn: (nil)
+    |   WriteTransferData: (nil) (nil)
+    |   Buffers:
+    |
+    | AxisData[0]:
+    |   Shape: 5
+    |   Index: 0
+    |   Strides: 16 8
+    |   Ptrs:
+    | AxisData[1]:
+    |   Shape: 10
+    |   Index: 0
+    |   Strides: 80 0
+    |   Ptrs:
+    ------- END ITERATOR DUMP -------
+    """).strip()
+
+    code = textwrap.dedent("""
+        import numpy as np
+        arr1 = np.arange(100, dtype=np.int64).reshape(10, 10)[:, ::2]
+        arr2 = np.arange(5.)
+        it = np.nditer((arr1, arr2), op_dtypes=["d", "i4"], casting="unsafe",
+                       flags=["reduce_ok", "buffered"],
+                       op_flags=[["readonly"], ["readwrite"]])
+        it.debug_print()
+        """)
+    res = subprocess.check_output([sys.executable, "-c", code], text=True)
+    res = res.strip()
+
+    for res_line, expected_line in zip(res.splitlines(), expected.splitlines()):
+        # The actual output may have additional pointers listed that are
+        # stripped from the example output:
+        assert res_line.startswith(expected_line)

--- a/numpy/core/tests/test_nditer.py
+++ b/numpy/core/tests/test_nditer.py
@@ -1963,10 +1963,45 @@ def test_iter_buffered_cast_structured_type():
     sdt2 = [('b', 'O'), ('a', 'f8')]
     a = np.array([(1, 2, 3), (4, 5, 6)], dtype=sdt1)
 
-    assert_raises(ValueError, lambda : (
-        nditer(a, ['buffered', 'refs_ok'], ['readwrite'],
-               casting='unsafe',
-               op_dtypes=sdt2)))
+    for intent in ["readwrite", "readonly", "writeonly"]:
+        # If the following assert fails, the place where the error is raised
+        # within nditer may change. That is fine, but it may make sense for
+        # a new (hard to design) test to replace it:
+        assert np.can_cast(a.dtype, sdt2, casting="unsafe")
+        simple_arr = np.array([1, 2], dtype="i")  # succeeds but needs clean up
+        with pytest.raises(ValueError):
+            nditer((simple_arr, a), ['buffered', 'refs_ok'], [intent, intent],
+                   casting='unsafe', op_dtypes=["f", sdt2])
+
+
+def test_buffered_cast_error_paths():
+    with pytest.raises(ValueError):
+        # The input is cast into an `S3` buffer
+        np.nditer((np.array("a", dtype="S1"),), op_dtypes=["i"],
+                  casting="unsafe", flags=["buffered"])
+
+    # The `M8[ns]` is cast into the `S3` output
+    it = np.nditer((np.array(1, dtype="i"),), op_dtypes=["S1"],
+                   op_flags=["writeonly"], casting="unsafe", flags=["buffered"])
+    with pytest.raises(ValueError):
+        with it:
+            buf = next(it)
+            buf[...] = "a"  # cannot be converted to int.
+
+    # The following gives an unraisable error, there are probably better
+    # ways to test that:
+    code = textwrap.dedent("""
+    import numpy as np
+
+    it = np.nditer((np.array(1, dtype="i"),), op_dtypes=["S1"],
+                   op_flags=["writeonly"], casting="unsafe", flags=["buffered"])
+    buf = next(it)
+    buf[...] = "a"
+    del buf, it  # Flushing only happens during deallocate right now.
+    """)
+    res = subprocess.check_output([sys.executable, "-c", code],
+                                  stderr=subprocess.STDOUT, text=True)
+    assert "ValueError" in res
 
 
 def test_iter_buffered_cast_subarray():

--- a/numpy/core/tests/test_nditer.py
+++ b/numpy/core/tests/test_nditer.py
@@ -1,7 +1,7 @@
 import sys
 import pytest
 
-import subprocess, textwrap, re
+import textwrap, subprocess
 
 import numpy as np
 import numpy.core._multiarray_tests as _multiarray_tests
@@ -1988,8 +1988,8 @@ def test_buffered_cast_error_paths():
             buf = next(it)
             buf[...] = "a"  # cannot be converted to int.
 
-    # The following gives an unraisable error, there are probably better
-    # ways to test that:
+    # The following gives an unraisable error. Pytest sometimes captures that
+    # (depending on version). So this can probably be cleaned out in the future:
     code = textwrap.dedent("""
     import numpy as np
 
@@ -3053,7 +3053,7 @@ def test_partial_iteration_error(in_dtype, buf_dtype):
     assert count == sys.getrefcount(value)
 
 
-def test_debug_print():
+def test_debug_print(capfd):
     """
     Matches the expected output of a debug print with the actual output.
     Note that the iterator dump should not be considered stable API,
@@ -3063,7 +3063,7 @@ def test_debug_print():
     """
     # the expected output with all addresses and sizes stripped (they vary
     # and/or are platform dependend).
-    expected = textwrap.dedent("""
+    expected = """
     ------ BEGIN ITERATOR DUMP ------
     | Iterator Address:
     | ItFlags: BUFFER REDUCE REUSE_REDUCE_LOOPS
@@ -3116,21 +3116,19 @@ def test_debug_print():
     |   Strides: 80 0
     |   Ptrs:
     ------- END ITERATOR DUMP -------
-    """).strip()
+    """.strip().splitlines()
 
-    code = textwrap.dedent("""
-        import numpy as np
-        arr1 = np.arange(100, dtype=np.int64).reshape(10, 10)[:, ::2]
-        arr2 = np.arange(5.)
-        it = np.nditer((arr1, arr2), op_dtypes=["d", "i4"], casting="unsafe",
-                       flags=["reduce_ok", "buffered"],
-                       op_flags=[["readonly"], ["readwrite"]])
-        it.debug_print()
-        """)
-    res = subprocess.check_output([sys.executable, "-c", code], text=True)
-    res = res.strip()
+    arr1 = np.arange(100, dtype=np.int64).reshape(10, 10)[:, ::2]
+    arr2 = np.arange(5.)
+    it = np.nditer((arr1, arr2), op_dtypes=["d", "i4"], casting="unsafe",
+                   flags=["reduce_ok", "buffered"],
+                   op_flags=[["readonly"], ["readwrite"]])
+    it.debug_print()
+    res = capfd.readouterr().out
+    res = res.strip().splitlines()
 
-    for res_line, expected_line in zip(res.splitlines(), expected.splitlines()):
+    assert len(res) == len(expected)
+    for res_line, expected_line in zip(res, expected):
         # The actual output may have additional pointers listed that are
         # stripped from the example output:
-        assert res_line.startswith(expected_line)
+        assert res_line.startswith(expected_line.strip())

